### PR TITLE
8321786: SegmentAllocator:allocateFrom(ValueLayout, MemorySegment,ValueLayout,long,long) spec mismatch in exception scenario

### PR DESCRIPTION
**Pull request by minborg**
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Change requires CSR request [JDK-8321923](https://bugs.openjdk.org/browse/JDK-8321923) to be approved
- [x] Commit message must refer to an issue

### Issues
 * [JDK-8321786](https://bugs.openjdk.org/browse/JDK-8321786): SegmentAllocator:allocateFrom(ValueLayout, MemorySegment,ValueLayout,long,long) spec mismatch in exception scenario (**Bug** - P3)
 * [JDK-8321923](https://bugs.openjdk.org/browse/JDK-8321923): SegmentAllocator:allocateFrom(ValueLayout, MemorySegment,ValueLayout,long,long) spec mismatch in exception scenario (**CSR**)

### Reviewers
 * [Maurizio Cimadamore](https://openjdk.org/census#mcimadamore) (@mcimadamore - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk22.git pull/42/head:pull/42` \
`$ git checkout pull/42`

Update a local copy of the PR: \
`$ git checkout pull/42` \
`$ git pull https://git.openjdk.org/jdk22.git pull/42/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 42`

View PR using the GUI difftool: \
`$ git pr show -t 42`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk22/pull/42.diff">https://git.openjdk.org/jdk22/pull/42.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk22/pull/42#issuecomment-1882592552)

### CODE REVIEW

File: src/java.base/share/classes/java/lang/foreign/MemorySegment.java [hunk 1]
```diff
@@ -1923,8 +1941,7 @@ static void copy(MemorySegment srcSegment, ValueLayout srcElementLayout, long sr
      *         <a href="MemorySegment.html#segment-alignment">incompatible with the alignment constraint</a>
      *         in the provided layout
      * @throws IndexOutOfBoundsException if {@code offset > byteSize() - layout.byteSize()}
-     * @throws UnsupportedOperationException if this segment is
-     *         {@linkplain #isReadOnly() read-only}
+     *         or {@code offset < 0}
```
Comment (mcimadamore) [inline src/java.base/share/classes/java/lang/foreign/MemorySegment.java:1944]: Why is UOE being removed here (and elsewhere) ?

Comment (minborg) [reply src/java.base/share/classes/java/lang/foreign/MemorySegment.java:1944]: This one is redundant and there is an IAE covering the same case a bit down.

File: test/jdk/java/foreign/TestSegmentAllocators.java [hunk 2]
```diff
@@ -189,6 +191,81 @@ public void testAllocatorAllocateFromHeapSegment() {
...
+                    arena.allocateFrom(elementLayout.withByteAlignment(elementLayout.byteAlignment() * 2), source, sourceElementLayout, 1, 1)
+            );
+
+            // IllegalStateException if the {@linkplain MemorySegment#scope() scope} associated
+            // with {@code source} is not {@linkplain MemorySegment.Scope#isAlive() alive}
+            // This is tested in TestScopedOperations
+
+            // WrongThreadException if this method is called from a thread {@code T},
+            // such that {@code source.isAccessibleBy(T) == false}
+            CompletableFuture<Arena> future = CompletableFuture.supplyAsync(Arena::ofConfined);
```
Comment (mcimadamore) [inline test/jdk/java/foreign/TestSegmentAllocators.java:234]: I believe TestScopedOperations should check for WTE as well?

Comment (mcimadamore) [reply test/jdk/java/foreign/TestSegmentAllocators.java:234]: That test has two methods: `testOpAfterClose` and `testOpOutsideConfinement`, which are executed on all the test cases.

**Review by mcimadamore — APPROVED** [review]

File: test/jdk/java/foreign/TestSegmentAllocators.java (hunk 2 above)
Comment (minborg) [reply test/jdk/java/foreign/TestSegmentAllocators.java:234]: Good idea. We could do that in a separate PR.

**Pull request closed without merging**